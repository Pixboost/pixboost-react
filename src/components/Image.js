import React, {Component} from 'react';
import PropTypes from 'prop-types';
import lozad from '../lozad';
import {isElementInViewport, sourceWithProtocol} from '../util';

class Image extends Component {
  constructor(props) {
    super(props);
    this.imgRef = React.createRef();
  }

  componentDidMount() {
    if (this.props.lazy) {
      const el = this.imgRef.current;
      const lozadInstance = lozad(el, {
        threshold: 0.01,
        rootMargin: '40px 0px 0px 0px',
      });
      if (isElementInViewport(el)) {
        lozadInstance.triggerLoad(el);
      } else {
        lozadInstance.observe();
      }
    }
  }

  render() {
    const {config, src, alt, op, lazy, ...otherProps} = this.props;
    const imgAttrs = {};
    const imgSrc = src.indexOf('data:') === 0 ?
      src :
      `${config.domain.includes('//') ? '' : '//'}${config.domain}/api/2/img/${sourceWithProtocol(src)}/${op}${op.includes('?') ? '&' : '?'}auth=${config.apiKey}`;

    if (lazy) {
      imgAttrs['data-src'] = imgSrc;
      imgAttrs['src'] = 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==';
    } else {
      imgAttrs['src'] = imgSrc;
    }
    if (alt) {
      imgAttrs.alt = alt;
    }

    return (
      <img {...imgAttrs} {...otherProps} ref={this.imgRef}/>
    );
  }
}

Image.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string,
  op: PropTypes.string,
  lazy: PropTypes.bool,
  config: PropTypes.object.isRequired
};

Image.defaultProps = {
  lazy: true,
  op: 'optimise'
};

export default Image;