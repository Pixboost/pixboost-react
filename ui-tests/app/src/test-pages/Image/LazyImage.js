import React, { Component } from 'react';
import {Image} from 'pixboost-react';
import {pixboostConfig} from '../config';

class LazyImage extends Component {
  render() {
    return (
      <div>
        <div style={{minHeight: '2000px'}}/>
        <Image src={'http://www.midday.coffee/assets/banner.jpg'} lazy={true} config={pixboostConfig}/>
      </div>
    );
  }
}

export default LazyImage;