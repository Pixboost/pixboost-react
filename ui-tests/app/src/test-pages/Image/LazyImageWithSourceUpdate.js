import React, { Component } from 'react';
import {Image} from 'pixboost-react';
import {pixboostConfig} from '../config';

class LazyImageWithSourceUpdate extends Component {
  constructor(props) {
    super(props);

    this.state = {
      src: 'http://www.midday.coffee/assets/cup.jpeg'
    }
  }

  render() {
    return (
      <div>
        <Image src={this.state.src} lazy={true} config={pixboostConfig}/>
        <button onClick={() => {this.setState({src: 'http://www.midday.coffee/assets/banner.jpg'})}}>Update source</button>
      </div>
    );
  }
}

export default LazyImageWithSourceUpdate;