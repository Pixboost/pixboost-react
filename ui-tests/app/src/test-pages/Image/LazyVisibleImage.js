import React, { Component } from 'react';
import {Image} from 'pixboost-react';
import {pixboostConfig} from '../config';

class LazyVisibleImage extends Component {
  render() {
    return (
      <div>
        <Image src={'http://www.midday.coffee/assets/banner.jpg'} lazy={true} config={pixboostConfig}/>
      </div>
    );
  }
}

export default LazyVisibleImage;