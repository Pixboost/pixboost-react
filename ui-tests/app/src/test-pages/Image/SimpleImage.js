import React, { Component } from 'react';
import {Image} from 'pixboost-react';
import {pixboostConfig} from '../config';

class SimpleImage extends Component {
  render() {
    const config = {
      apiKey: 'MTg4MjMxMzM3MA__',
      domain: 'pixboost.com'
    };
    return (
      <Image src={'http://www.midday.coffee/assets/banner.jpg'} lazy={false} config={pixboostConfig}/>
    );
  }
}

export default SimpleImage;