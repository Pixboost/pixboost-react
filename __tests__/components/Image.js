import Image from '../../src/components/Image';
import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme/build';

const mockLozad = {
  observe: jest.fn(),
  triggerLoad: jest.fn()
};

jest.mock('../../src/lozad', () => {
  return jest.fn().mockImplementation(() => {
    return mockLozad;
  });
});

const testRendererOptions = {
  createNodeMock: () => {
    return {
      getBoundingClientRect: () => { return {}; }
    };
  }
};

const testConfig = {
  domain: 'https://test.com',
  apiKey: 'abc123'
};

describe('Image', () => {
  beforeEach(() => {
    mockLozad.observe.mockClear();
    mockLozad.triggerLoad.mockClear();
  });

  it('should not crash when dont have expected parameters', () => {
    renderer.create(<Image src={undefined} config={undefined}/>, testRendererOptions);
  });

  it('should render with minimum parameters', () => {
    expect(
      renderer.create(<Image src={'https://image.here.com/logo.png'} config={testConfig}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

  it('should render with domain that don\'t have scheme', () => {
    expect(
      renderer.create(<Image src={'https://image.here.com/logo.png'} config={{domain: 'test.com', apiKey: 'abc123'}}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

  it('should render with alt text', () => {
    expect(
      renderer.create(<Image src={'https://image.here.com/logo.png'} config={testConfig} alt={'yo!'}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

  it('should render with lazy loading disabled', () => {
    expect(
      renderer.create(<Image src={'https://image.here.com/logo.png'} config={testConfig} lazy={false}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

  it('should render with operation that have arguments', () => {
    expect(
      renderer.create(<Image src={'https://image.here.com/logo.png'} config={testConfig} op={'fit?size=100x200'}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

  it('should not apply URL transformation for data: sources', () => {
    expect(
      renderer.create(<Image src={'data:ABCDEF'} config={testConfig} op={'fit?size=100x200'}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

  it('should replace source that starts with // to https://', () => {
    expect(
      renderer.create(<Image src={'//image.here.com/logo.png'} config={testConfig} op={'fit?size=100x200'}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

  it('should setup lazy loading when props changed', () => {
    const image = mount(<Image src={'//image.here.com/logo.png'} config={testConfig} op={'fit?size=100x200'}/>);
    expect(mockLozad.triggerLoad).toHaveBeenCalledTimes(1);

    image.setProps({
      src: '//newimage.here.com'
    });

    expect(mockLozad.triggerLoad).toHaveBeenCalledTimes(2);
  });

  it('should encode image source if it has query params', () => {
    expect(
      renderer.create(<Image src={'//image.here.com/logo.png?param=1'} config={testConfig} op={'fit?size=100x200'}/>, testRendererOptions).toJSON()
    ).toMatchSnapshot();
  });

});