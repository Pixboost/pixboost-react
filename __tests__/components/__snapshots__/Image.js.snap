// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should encode image source if it has query params 1`] = `
<img
  data-src="https://test.com/api/2/img/http:%2F%2Fimage.here.com%2Flogo.png%3Fparam%3D1/fit?size=100x200&auth=abc123"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
/>
`;

exports[`Image should not apply URL transformation for data: sources 1`] = `
<img
  data-src="data:ABCDEF"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
/>
`;

exports[`Image should render with alt text 1`] = `
<img
  alt="yo!"
  data-src="https://test.com/api/2/img/https://image.here.com/logo.png/optimise?auth=abc123"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
/>
`;

exports[`Image should render with domain that don't have scheme 1`] = `
<img
  data-src="//test.com/api/2/img/https://image.here.com/logo.png/optimise?auth=abc123"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
/>
`;

exports[`Image should render with lazy loading disabled 1`] = `
<img
  src="https://test.com/api/2/img/https://image.here.com/logo.png/optimise?auth=abc123"
/>
`;

exports[`Image should render with minimum parameters 1`] = `
<img
  data-src="https://test.com/api/2/img/https://image.here.com/logo.png/optimise?auth=abc123"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
/>
`;

exports[`Image should render with operation that have arguments 1`] = `
<img
  data-src="https://test.com/api/2/img/https://image.here.com/logo.png/fit?size=100x200&auth=abc123"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
/>
`;

exports[`Image should replace source that starts with // to https:// 1`] = `
<img
  data-src="https://test.com/api/2/img/http://image.here.com/logo.png/fit?size=100x200&auth=abc123"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
/>
`;
